crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,738 @@
+            second_instr.params.iter().any(|p| !matches!(p, Param::Float(_)));
+
+        if skip_cache {
+            return self.commute_matmul(first_instr, first_qargs, second_instr, second_qargs);
+        }
+
+        // Query commutation library
+        if let Some(is_commuting) =
+            self.library
+                .check_commutation_entries(&first_op, first_qargs, &second_op, second_qargs)
+        {
+            return is_commuting;
+        }
+        // Query cache
+        if let Some(commutation_dict) = self
+            .cache
+            .get(&(first_op.name().to_string(), second_op.name().to_string()))
+        {
+            if let Some(commutation) = commutation_dict.get(&(
+                get_relative_placement(first_qargs, second_qargs),
+                (
+                    hashable_params(&first_instr.params),
+                    hashable_params(&second_instr.params),
+                ),
+            )) {
+                self._cache_hit += 1;
+                return *commutation;
+            } else {
+                self._cache_miss += 1;
+            }
+        } else {
+            self._cache_miss += 1;
+        }
+
+        // Perform matrix multiplication to determine commutation
+        let is_commuting =
+            self.commute_matmul(first_instr, first_qargs, second_instr, second_qargs);
+
+        // TODO: implement a LRU cache for this
+        if self.current_cache_entries >= self.cache_max_entries {
+            self.clear_cache();
+        }
+        // Cache results from is_commuting
+        self.cache
+            .entry((first_op.name().to_string(), second_op.name().to_string()))
+            .and_modify(|entries| {
+                let key = (
+                    get_relative_placement(first_qargs, second_qargs),
+                    (
+                        hashable_params(&first_instr.params),
+                        hashable_params(&second_instr.params),
+                    ),
+                );
+                entries.insert(key, is_commuting);
+                self.current_cache_entries += 1;
+            })
+            .or_insert_with(|| {
+                let mut entries = HashMap::with_capacity(1);
+                let key = (
+                    get_relative_placement(first_qargs, second_qargs),
+                    (
+                        hashable_params(&first_instr.params),
+                        hashable_params(&second_instr.params),
+                    ),
+                );
+                entries.insert(key, is_commuting);
+                self.current_cache_entries += 1;
+                CommutationCacheEntry { mapping: entries }
+            });
+        is_commuting
+    }
+
+    fn commute_matmul(
+        &self,
+        first_instr: &CircuitInstruction,
+        first_qargs: &[usize],
+        second_instr: &CircuitInstruction,
+        second_qargs: &[usize],
+    ) -> bool {
```

Comment:
Same here, could we use `u32` instead?
```suggestion
    fn commute_matmul(
        &self,
        first_instr: &CircuitInstruction,
        first_qargs: &[u32],
        second_instr: &CircuitInstruction,
        second_qargs: &[u32],
    ) -> bool {
```